[Core-9143] pandaproxy/sr: Add support for serialized format option in protobuf

IoannisRP opened the pull request:
Implements: [CORE-9143](https://redpandadata.atlassian.net/browse/CORE-9143)

Also acts as a followup to #25971 where:
- missing api specification is added
- support for format query parameter is added to endpoints:
  - `post subject/{sub}`
  - `get subject/{subject}/version/{version}`
  - `get subject/{subject}/version/{version}/schema`
- relax error handling
- change error thrown from `400` -> `501`

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

### Features

- Added support for the `format=serialized` query parameter for protobuf schemas for the following schema registry endpoints:
  - `get schema/ids/{id}`
  - `post subject/{sub}`
  - `get subject/{subject}/version/{version}`
  - `get subject/{subject}/version/{version}/schema`


[CORE-9143]: https://redpandadata.atlassian.net/browse/CORE-9143?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

IoannisRP committed:
pandaproxy/sr: Add support for serialized format value in protobuf

IoannisRP commented on src/v/pandaproxy/schema_registry/test/protobuf_rendering.cc:
```diff
@@ -60,18 +60,47 @@ struct simple_sharded_store {
+
+enum class SchemaType { input, sanitized, normalized };
+
+std::string_view to_string_view(const SchemaType st) {
+    switch (st) {
+    case SchemaType::input:
+        return "input";
+    case SchemaType::sanitized:
+        return "sanitized";
+    case SchemaType::normalized:
+        return "normalized";
+    }
+}
+
+ss::sstring read_schema(const std::string_view test_case, const SchemaType pt) {
+    const auto proto_file = fmt::format(
+      "{}_{}.proto", test_case, to_string_view(pt));
```
Fun fact:
`fmt` is not meant to be extendable by `to_string_view`, however, it does currently extend the serialization to this type. I.e. I could remove the `to_string_view` call in the `fmt::format` and it still works.
However, as it is a side-effect and not by design, i preferred to be more explicit.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--71385a89-74fc-494e-9051-85a3cdab1911-->
<!--TEST RESULTS-->
<!--9ad94d1d-830f-4fd8-b6cc-bd74bb60d07b-->
<!--TEST RESULTS-->
<!--0c7e9aae-587d-4398-b479-65830f4ecbed-->
<!--TEST RESULTS-->
<!--bce0b13b-6bc8-4c5c-b4c7-5740b9b6d6c1-->
<!--TEST RESULTS-->
<!--7512b5cc-27a0-4e41-b7e3-004639ef7866-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65913">build#65913</a></summary>


| test_class                 | test_method                                       | test_arguments                                                                                                                                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:---------------------------|:--------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest      | test_migrated_topic_data_integrity                | {"params": {"cancellation": {"dir": "out", "stage": "prepared"}, "use_alias": false}, "transfer_leadership": true}                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65913#0196ca62-83ba-4979-8f6e-649036bb8876 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| PartitionReassignmentsTest | test_add_partitions_with_inprogress_reassignments |                                                                                                                                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65913#0196ca61-d305-49f7-bbb9-fad34e698f2f | FLAKY         | 18/21    | upstream reliability is '89.0595009596929'. current run reliability is '85.71428571428571'. drift is 3.34522 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest   | test_node_operations                              | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": false, "with_tiered_storage": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65913#0196ca61-d304-400e-b318-d20230d6e13d | FLAKY         | 20/21    | upstream reliability is '99.16666666666667'. current run reliability is '95.23809523809523'. drift is 3.92857 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest   | test_node_operations                              | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": true, "with_tiered_storage": false}  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65913#0196ca61-d304-400e-b318-d20230d6e13d | FLAKY         | 20/21    | upstream reliability is '96.16519174041298'. current run reliability is '95.23809523809523'. drift is 0.9271 and the allowed drift is set to 50. The test should PASS  |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/65979">build#65979</a></summary>


| test_class                | test_method                     | test_arguments                                                                                                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:--------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageChunkReadTest | test_read_chunks                |                                                                                                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65979#0196ce56-2dc5-4329-9ff0-9cb3828ebe10 | FLAKY         | 20/21    | upstream reliability is '99.54337899543378'. current run reliability is '95.23809523809523'. drift is 4.30528 and the allowed drift is set to 50. The test should PASS |
| DeleteRecordsKafkaTest    | test_delete_records_empty_topic |                                                                                                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65979#0196ce53-d769-4f7e-8854-0f6febbd9219 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
| OffsetForLeaderEpochTest  | test_offset_for_leader_epoch    |                                                                                                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65979#0196ce56-2dc6-434a-bb14-8e7cbe71c2cb | FLAKY         | 20/21    | upstream reliability is '99.54337899543378'. current run reliability is '95.23809523809523'. drift is 4.30528 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest  | test_node_operations            | {"cloud_storage_type": 1, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": true, "with_tiered_storage": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65979#0196ce53-d769-4f7e-8854-0f6febbd9219 | FLAKY         | 20/21    | upstream reliability is '95.53429027113238'. current run reliability is '95.23809523809523'. drift is 0.2962 and the allowed drift is set to 50. The test should PASS  |
| JavaClientOIDCTest        | test_java_client                |                                                                                                                                                                   | ducktape    | https://buildkite.com/redpanda/redpanda/builds/65979#0196ce53-d769-4f7e-8854-0f6febbd9219 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66065">build#66065</a></summary>


| test_class            | test_method                        | test_arguments                                                                                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:----------------------|:-----------------------------------|:-------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| DataMigrationsApiTest | test_migrated_topic_data_integrity | {"params": {"cancellation": {"dir": "in", "stage": "executing"}, "use_alias": false}, "transfer_leadership": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66065#0196d518-51c8-4fc9-bc3f-a752f7a3b290 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66115">build#66115</a></summary>


| test_class                | test_method                           | test_arguments                                                                                                                                                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:--------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterHealthOverviewTest | cluster_health_overview_baseline_test |                                                                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66115#0196d9fc-78df-4907-a814-d4e4862269d5 | FLAKY         | 19/21    | upstream reliability is '95.88235294117648'. current run reliability is '90.47619047619048'. drift is 5.40616 and the allowed drift is set to 50. The test should PASS |
| DeleteRecordsKafkaTest    | test_delete_records_empty_topic       |                                                                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66115#0196d9e8-6d9e-41c7-882c-2a483ce445ee | FLAKY         | 20/21    | upstream reliability is '99.10313901345292'. current run reliability is '95.23809523809523'. drift is 3.86504 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest  | test_node_operations                  | {"cloud_storage_type": 2, "enable_failures": false, "mixed_versions": false, "with_chunked_compaction": false, "with_iceberg": false, "with_tiered_storage": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66115#0196d9fc-78df-4907-a814-d4e4862269d5 | FLAKY         | 20/21    | upstream reliability is '98.08917197452229'. current run reliability is '95.23809523809523'. drift is 2.85108 and the allowed drift is set to 50. The test should PASS |
| JavaClientOIDCTest        | test_java_client                      |                                                                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66115#0196d9e8-6d9e-41c7-882c-2a483ce445ee | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS              |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/66270">build#66270</a></summary>


| test_class               | test_method                    | test_arguments                                                                                                                      | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:-------------------------|:-------------------------------|:------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| AuditLogTestOauth        | test_kafka_oauth               | {"authz_match": "acl"}                                                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66270#0196f417-72d8-4f61-a660-7f68770e3196 | FLAKY         | 20/21    | upstream reliability is '97.49430523917995'. current run reliability is '95.23809523809523'. drift is 2.25621 and the allowed drift is set to 50. The test should PASS  |
| STSRoleFetchTests        | test_write                     |                                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66270#0196f417-72d8-4f61-a660-7f68770e3196 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
| PartitionBalancerTest    | test_fuzz_admin_ops            |                                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66270#0196f411-6200-4cff-90dc-066db5a51786 | FLAKY         | 17/21    | upstream reliability is '93.46938775510203'. current run reliability is '80.95238095238095'. drift is 12.51701 and the allowed drift is set to 50. The test should PASS |
| RaftAvailabilityTest     | test_controller_node_isolation |                                                                                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66270#0196f411-6201-4821-a817-add91a6c97b0 | FLAKY         | 20/21    | upstream reliability is '97.48953974895397'. current run reliability is '95.23809523809523'. drift is 2.25144 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest | test_node_operations           | {"cloud_storage_type": 2, "enable_failures": false, "mixed_versions": true, "with_chunked_compaction": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66270#0196f411-61fe-4a8f-87b0-3811404af561 | FLAKY         | 20/21    |                                                                                                                                                                         |
| RandomNodeOperationsTest | test_node_operations           | {"cloud_storage_type": 2, "enable_failures": true, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/66270#0196f411-61fe-4a8f-87b0-3811404af561 | FLAKY         | 20/21    |                                                                                                                                                                         |
</details>
<!--/body-->

michael-redpanda commented on src/v/pandaproxy/schema_registry/sharded_store.cc:
```diff
@@ -376,15 +376,26 @@ sharded_store::get_schema_definition(schema_id id) {
           return s.get_schema_definition(id).value();
       });
+
+    switch (format) {
+    case output_format::none:
+        co_return def;
+    case output_format::resolved:
+    case output_format::ignore_extensions:
+        throw as_exception(format_not_supported(format));
+    case output_format::serialized:
+        if (def.type() != schema_type::protobuf) {
+            throw std::runtime_error(
+              fmt::format("Bad format '{}' for type '{}'", format, def.type()));
+        }
+        auto serialized = co_await make_canonical_protobuf_schema(
+          *this, {{}, std::move(def)}, normalize::no, format);
+        co_return std::move(serialized).def();
```
comment/suggestion: What about throwing in the implementation?  I'm wondering if checking for valid format makes more sense to occur in the schema type implementation

IoannisRP replied:
In general, `sharded_store` is both type and args aware. For example, `make_canonical_schema` looks like this:
https://github.com/redpanda-data/redpanda/blob/8ae76ce365f85c7efab09e8cf970960f6848195e/src/v/pandaproxy/schema_registry/sharded_store.cc#L82-L103
A switch that behaves differently based on schema type is used across quite a bit of its functionality.

Alternatively, I could add a `format_schema` across the 3 types (like the 3 different versions of `make_canonical_XXX_schema` for each type) and just forward `def` into that. Minor detail, this would have extra move (or 2), for the default case of `output_format::none`.
Would you prefer such an approach?

michael-redpanda commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -333,13 +333,22 @@ get_schemas_ids_id(server::request_t rq, server::reply_t rp) {
     parse_accept_header(rq, rp);
     auto id = parse::request_param<schema_id>(*rq.req, "id");
 
+    const auto format_str = parse::query_param<std::optional<ss::sstring>>(
```
I'm not seeing any changes for;
* POST subjects subject
* POST subjects subject versions
* GET subjects/subject/versions/version
*

IoannisRP replied:
Yes, these endpoints were not part of the discussed changes in the rfc. Happy to discuss this offline.

michael-redpanda replied:
I think if we're adding support for `format` in one endpoint, it makes sense to add it for others.  The lift shouldn't be too bad right?

michael-redpanda reviewed (changes requested):
Are some APIs missing?

BenPope commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -333,13 +333,22 @@ get_schemas_ids_id(server::request_t rq, server::reply_t rp) {
+                              .value_or("");
+    const auto format = from_string_view<output_format>(format_str);
+    if (!format.has_value()) {
+        throw as_exception(invalid_format(format_str));
```
This seems stricter than the reference implementation.

BenPope commented on src/v/pandaproxy/schema_registry/protobuf.cc:
```diff
@@ -546,13 +546,21 @@ struct protobuf_schema_definition::impl {
+            fdp.SerializeToOstream(&ios.ostream());
+
+            return schema_definition::raw_string{
+              iobuf_to_base64(std::move(ios).buf())};
```
We should implement an `iobuf_to_base64_iobuf` to prevent a linearisation.

IoannisRP replied:
I agree, but given that this would be a base64 util, i would prefer to keep it separate to this PR

BenPope replied:
Is there a ticket for this work? Would it make sense to commit that PR first, and then base this on that one?

I'm not opposed to including it in this PR.

BenPope commented on src/v/pandaproxy/schema_registry/protobuf.cc:
```diff
@@ -634,18 +642,25 @@ ss::future<protobuf_schema_definition> make_protobuf_schema_definition(
     auto res = co_await make_protobuf_schema_definition(
       store, std::move(schema), norm);
-    co_return schema_definition{std::move(res)};
+    co_return schema_definition{res.raw(format), res.type(), res.refs()};
```
Is this a bug fix? Can it go in a separate commit with a test.

IoannisRP replied:
No, not a bug fix. Type of `res` is `protobuf_schema_definition` which has an implicit conversion operator to `schema_definition`
https://github.com/redpanda-data/redpanda/blob/8ae76ce365f85c7efab09e8cf970960f6848195e/src/v/pandaproxy/schema_registry/types.h#L275-L277
Since implicit conversions don't take parameters, i had to break it up

BenPope commented on src/v/pandaproxy/schema_registry/sharded_store.cc:
```diff
@@ -376,15 +376,26 @@ sharded_store::get_schema_definition(schema_id id) {
+        throw as_exception(format_not_supported(format));
+    case output_format::serialized:
+        if (def.type() != schema_type::protobuf) {
+            throw std::runtime_error(
```
We don't usually throw `std::runtime_error`, why not `format_not_supported` or `invalid_format`, but I'm not sure rejecting is compatible with the reference implementation.

IoannisRP replied:
This seems like a leftover from initial implementation. I'll update it to `invalid_format` while we discuss discrepancies with reference implementation...

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/4e03ff1dd56e44b865db8e68a93f05d9ce60f35a..b8617d2f60d4bdeedadcc0029bc467f8eff8adb9):
- rebase to dev

IoannisRP committed:
pandaproxy/sr: Relax error handling around format argument

- Replace throwing logic with default behavior
- Remove dead code

IoannisRP committed:
pandaproxy/sr: push format handling down into the separate types

- Push the behavior from sharded store into the separate types.
- Further relax the error handling by throwing only when a format type
is called on the type it would affect.

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/b8617d2f60d4bdeedadcc0029bc467f8eff8adb9..bfc68af6661e68c524cb1638f8af1c6f884f6283):
- Relax error handling
- Remove dead code
- Encapsulate the format behavior within each type

Note: `format` is still only enabled for `/schemas/ids/{id}`

IoannisRP committed:
pandaproxy/sr: Fix missing schema registry api definition

IoannisRP committed:
pandaproxy/sr: Introduce format parameter to post subjects subject endpoint

IoannisRP committed:
pandaproxy/sr: Introduce format parameter to get subjects subject versions version endpoint

IoannisRP committed:
pandaproxy/sr: Introduce format parameter to get subjects subject versions version schema endpoint

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/bfc68af6661e68c524cb1638f8af1c6f884f6283..44c8eb35dbc25df75920aa322899596a9880571d):
- add missing query parameter for `/schemas/ids/{id}` in `schema_registry.json`
- add `format` support in endpoints
  * `POST /subjects/{subject}`
  * `GET /subjects/{subject}/versions/{version}`
  * `GET /subjects/{subject}/versions/{version}/schema`

BenPope commented on tests/rptest/tests/schema_registry_test.py:
```diff
@@ -3626,14 +3631,26 @@ def test_ids_id(id, schema, format=None):
+                                                     schema_proto_b64,
+                                                     "schemaType":
+                                                     "PROTOBUF"
+                                                 }))
```
Support for this isn't implemented until the next commit.

IoannisRP replied:
>The serialized output seems to be longer than the reference implementation.

I'll have a look...

>Are tests needed against the javascript client?

As long as we are compatible with reference API, i don't think so, but open to what other people feel.

> Support for this isn't implemented until the next commit.

This is to test that we accept schemas in b64. While technically not a new feature, i didn't see any other tests for it. Also it is a good test to ensure that our `serialized` output is a valid schema and not a random b64 string.
In the next commit, support for the `serialized` keyword is implemented, which means that the schema will be returned in b64. Both raw test and b64 are valid input to that operation.

IoannisRP replied:
> The serialized output seems to be longer than the reference implementation.

Our implementation maintains custom options and respects round trips (posting text, receiving b64, reposting b64 and getting stored schema). Specifically, we can do lookup by b64 with our implementation but it fails with b64 provided by reference.

Reference implementation doesn't maintain custom options in b64. So, if the schema has custom options, lookup by b64 always fails. Also, using b64 from our implementation also fails. I would assume custom options are not parsed, but this is guesswork.

In the absence of custom options, ours and reference b64 different but compatible (tested with normalization turned on).

BenPope replied:
> > Support for this isn't implemented until the next commit.

> This is to test that we accept schemas in b64. 

You are correct, not sure what I was thinking.

BenPope reviewed (commented):
Looks OK to me.

I'd rather the refactoring was done earlier, and in isolation, and then the rest of the commits building on it.

The serialized output seems to be longer than the reference implementation.

Are tests needed against the javascript client?

michael-redpanda commented on src/v/pandaproxy/schema_registry/sharded_store.cc:
```diff
@@ -118,6 +119,21 @@ ss::future<> sharded_store::validate_schema(subject_schema schema) {
     __builtin_unreachable();
 }
 
+ss::future<schema_definition>
+sharded_store::format_schema(schema_definition schema, output_format format) {
+    switch (schema.type()) {
+    case schema_type::avro:
+        co_return co_await format_avro_schema_definition(
+          *this, std::move(schema), format);
+    case schema_type::protobuf:
+        co_return co_await format_protobuf_schema_definition(
+          *this, std::move(schema), format);
+    case schema_type::json:
+        // json doesn't act on format parameter
+        co_return schema;
+    }
+}
+
```
chore: this was confusing me because I thought you had added it earlier but you just moved it in the same PR... is there any refactoring of the commit history we can do to clean up the history?  It's a little confusing to follow

IoannisRP replied:
The behavior that split the error handling between the types comes from this PR. Previously it was just checking to see if it is not `output_format::none`

michael-redpanda commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -75,7 +75,10 @@ parse_schema_version(const ss::sstring& ver) {
 }
 
-output_format parse_output_format(const ss::sstring& format_str) {
+output_format parse_output_format(const ss::http::request& req) {
```
The changes here really should be built upon in future commits rather than refactoring in the middle of the PR.  We should be following the guidance in our contributing.md file.

IoannisRP replied:
i'll clean up the history and split it into 2 PRs

BenPope replied:
> split it into 2 PRs

I don't think it's helpful to split it into 2 PRs at this point.

michael-redpanda reviewed (changes requested):
The changes look good, but the commit history is a bit messy.  Can it please be cleaned up

IoannisRP committed:
pandaproxy/sr: add utility function to read format query parameter

IoannisRP committed:
pandaproxy/sr: rework format test to be more scalable

IoannisRP commented:
Changes in [force-push](https://github.com/redpanda-data/redpanda/compare/44c8eb35dbc25df75920aa322899596a9880571d..449c139133301e227eedb2fcb57f5af43090e901):
- Reworked commit history. All the post-work/refactoring happens first and then the protobuf serialization is implemented
- minor update to error reporting and util function in dt test

CHanges in [force-push](https://github.com/redpanda-data/redpanda/compare/449c139133301e227eedb2fcb57f5af43090e901..9675dfef701fd6e785d1dca72905ee0c3b108c8a):
- Fix leftover from reworking history

BenPope commented on src/v/pandaproxy/schema_registry/handlers.cc:
```diff
@@ -75,6 +75,14 @@ parse_schema_version(const ss::sstring& ver) {
+    const auto format_str = parse::query_param<std::optional<ss::sstring>>(
+                              req, "format")
+                              .value_or("");
+    return from_string_view<output_format>(format_str)
```
Nitpick (just a PSA):
```suggestion
    return parse::query_param<std::optional<ss::sstring>>(req, "format")
      .and_then(&from_string_view<output_format>)
```

IoannisRP replied:
`1.0.5` -> `1.0.6`? it is backwards compatible

BenPope reviewed (commented):
It's probably worth bumping the API version in `src/v/pandaproxy/api/api-doc/schema_registry_header.json`

IoannisRP committed:
pandaproxy/sr: change error for not implemented format query parameters

- change error from 400 to 501
- update sr dt infra to not retry 501 errors

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/9675dfef701fd6e785d1dca72905ee0c3b108c8a..e003bdb0adfd3cd0274be20c820c4900fe949ff1):
- Change error type from `400` -> `501` for unimplemented format values

IoannisRP committed:
pandaproxy/sr: update version of sr api version

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/e003bdb0adfd3cd0274be20c820c4900fe949ff1..906c93b0a3fa7f42565072147237f75d8f939b3c):
- update api version
- address nit

michael-redpanda commented on src/v/pandaproxy/schema_registry/error.cc:
```diff
@@ -140,7 +140,7 @@ struct error_category final : std::error_category {
             return reply_error_code::internal_server_error; // 500
         case error_code::format_not_supported:
-            return reply_error_code::bad_request; // 400
+            return reply_error_code::not_implemented; // 501
```
nice

michael-redpanda reviewed (approved):
lgtm, thanks for cleaning it up!
